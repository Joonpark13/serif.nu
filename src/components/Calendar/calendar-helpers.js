export function getHours() {
  const hours = [];
  for (let i = 8; i <= 22; i++) {
    hours.push(i);
  }
  return hours;
}

export function parseHour(timeStr) {
  // timeStr format: 3:00PM or 11:50AM
  const hour = parseInt(timeStr.split(':')[0], 10);
  const ampm = timeStr.slice(-2);
  if (ampm === 'AM') {
    if (hour === 12) return 0; // 12AM
    return hour; // All other AMs
  }
  if (hour === 12) return hour; // 12PM
  return hour + 12; // All other PMs
}

export function parseMinute(timeStr) {
  // timeStr format: 3:00PM or 11:50AM
  return parseInt(timeStr.split(':')[1].slice(0, 2), 10);
}

export function parseMeetingTime(meetingTime) {
  // meetingTime format: MoWe 11:00AM - 11:50AM or MoWeFr 3:00PM - 3:50PM
  // If class is unscheduled, meetingTime will be 'TBA'
  if (meetingTime === 'TBA') return meetingTime;

  const splitStr = meetingTime.split(' ');
  const dow = splitStr[0];
  const start = splitStr[1];
  // the dash is splitStr[2]
  const end = splitStr[3];

  return {
    dow: dow.match(/.{2}/g), // Array of dow str broken up every two chars, ex: ['Mo', 'We']
    start: {
      hour: parseHour(start),
      minute: parseMinute(start),
    },
    end: {
      hour: parseHour(end),
      minute: parseMinute(end),
    },
  };
}
