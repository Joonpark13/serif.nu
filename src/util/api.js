/* eslint-disable import/prefer-default-export */
export function searchURL(query) {
  return `${process.env.API_URL}/search?term_id=4720&search_query=${query}`;
}

export function sectionsURL(termId, schoolAbbv, subjectAbbv, courseAbbv) {
  return `${process.env.API_URL}/sections?term_id=${termId}&school_abbv=${schoolAbbv}&subject_abbv=${subjectAbbv}&course_abbv=${courseAbbv}`;
}
