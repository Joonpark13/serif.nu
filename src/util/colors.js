export const northwesternPurple = '#4E2A84';
export const northwesternBrightOrange = '#FFC520';
export const northwesternRichBlack10 = '#D8D6D6';
