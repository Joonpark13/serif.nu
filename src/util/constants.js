/* eslint-disable import/prefer-default-export */
export const CURRENT_TERM = '4730';
