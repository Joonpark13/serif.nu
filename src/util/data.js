import termsData from 'data/terms.json';

/* eslint-disable import/prefer-default-export */
/* istanbul ignore next */
export const currentTerm = termsData.reduce((highestTerm, term) => {
  if (term.id > highestTerm.id) {
    return term;
  }
  return highestTerm;
});
/* eslint-enable import/prefer-default-export */
