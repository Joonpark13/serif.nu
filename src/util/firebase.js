import firebase from 'firebase/app';
import 'firebase/firestore';
import 'firebase/auth';

firebase.initializeApp({
  apiKey: 'AIzaSyDluAG6bnqleqJ3yjPvGdlb3RyZP3U5JIQ',
  authDomain: 'sans-serif-northwestern.firebaseapp.com',
  databaseURL: 'https://sans-serif-northwestern.firebaseio.com',
  projectId: 'sans-serif-northwestern',
  storageBucket: 'sans-serif-northwestern.appspot.com',
  messagingSenderId: '1078100635889',
});

export const db = firebase.firestore();
export const auth = firebase.auth();
