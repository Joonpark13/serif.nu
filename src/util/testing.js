/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import * as reactRedux from 'react-redux';
import { shallow } from 'enzyme';
import * as useSelector from './use-selector';

jest.mock('./use-selector');

// Helper function that returns a function which creates enzyme wrappers with default props
export function wrapperCreator(Component, defaultProps = {}) {
  let classes;
  return props => shallow(<Component classes={classes} {...defaultProps} {...props} />);
}

/* istanbul ignore next */
export function mockUseSelector(...args) {
  useSelector.default.mockReset();
  args.forEach(arg => useSelector.default.mockReturnValueOnce(arg));

  return useSelector.default;
}

/* istanbul ignore next */
export function mockUseDispatch() {
  const dispatchSpy = jest.fn();
  reactRedux.useDispatch.mockReturnValue(dispatchSpy);
  return dispatchSpy;
}
