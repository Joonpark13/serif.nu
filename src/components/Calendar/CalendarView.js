import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import { getHours, parseMeetingTime } from './calendar-helpers';
import HoursColumn from './HoursColumn';
import DowColumn from './DowColumn';

export const styles = {
  calendarRoot: {
    height: 'calc(100vh - 64px)', // 64px is height of TopBar
    display: 'flex',
  },
  dowColumns: {
    display: 'flex',
    width: '100%',
  },
};

const hours = getHours();
const dows = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri'];


export function getSectionsForDow(dow, sections) {
  // sections: [{
  //   class_mtg_info: [{
  //     meet_t: MoWe 11:00AM - 11:50AM
  //   }]
  // }]
  return sections.filter(section => section.class_mtg_info.some((mtgInfo) => {
    const mtgTime = parseMeetingTime(mtgInfo.meet_t);
    if (mtgTime === 'TBA') return false; // TODO: Handle unscheduled courses
    return mtgTime.dow.some(mtgTimeDow => dow.slice(0, 2) === mtgTimeDow);
    // mtgTimeDow will be 'Mo' and the dow param of the function will be 'Mon'
  }));
}

function CalendarView({ sections, classes }) {
  return (
    <div className={classes.calendarRoot}>
      <HoursColumn hours={hours} />

      <div className={classes.dowColumns}>
        {dows.map(dow => (
          <DowColumn key={dow} dow={dow} sections={getSectionsForDow(dow, sections)} />
        ))}
      </div>
    </div>
  );
}

CalendarView.propTypes = {
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
  sections: PropTypes.arrayOf(PropTypes.object).isRequired,
};

export { CalendarView as UnstyledCalendarView };

export default withStyles(styles)(CalendarView);
