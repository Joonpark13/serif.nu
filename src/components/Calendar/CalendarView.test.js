import React from 'react';
import { shallow } from 'enzyme';
import { mockStyles } from 'util/testing';
import * as calendarHelpers from './calendar-helpers';
import { getSectionsForDow, UnstyledCalendarView, styles } from './CalendarView';

describe('getSectionsForDow', () => {
  it('filters sections correctly given a dow', () => {
    const parseMeetingTimeSpy = jest.spyOn(calendarHelpers, 'parseMeetingTime');
    parseMeetingTimeSpy.mockReturnValue({
      dow: ['Mo', 'We'],
      start: {
        hour: 11,
        minute: 0,
      },
      end: {
        hour: 11,
        minute: 50,
      },
    });
    const sections = [{
      class_mtg_info: [{
        meet_t: 'MoWe 11:00AM - 11:50AM',
      }],
    }];

    expect(getSectionsForDow('Mon', sections)).toEqual(sections);
  });

  it('filter unscheduled sections correctly', () => {
    const parseMeetingTimeSpy = jest.spyOn(calendarHelpers, 'parseMeetingTime');
    parseMeetingTimeSpy.mockReturnValue('TBA');
    const sections = [{
      class_mtg_info: [{
        meet_t: 'TBA',
      }],
    }];

    expect(getSectionsForDow('Mon', sections)).toEqual([]);
  });
});

describe('CalendarView', () => {
  it('renders correctly', () => {
    const classes = mockStyles(styles);
    const wrapper = shallow(<UnstyledCalendarView sections={[]} classes={classes} />);

    expect(wrapper).toMatchSnapshot();
  });
});
