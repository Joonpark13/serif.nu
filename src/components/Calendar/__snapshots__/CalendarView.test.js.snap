// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarView renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CalendarView
    classes={
      Object {
        "calendarRoot": "",
        "dowColumns": "",
      }
    }
    sections={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <WithStyles(HoursColumn)
          hours={
            Array [
              8,
              9,
              10,
              11,
              12,
              13,
              14,
              15,
              16,
              17,
              18,
              19,
              20,
              21,
              22,
            ]
          }
        />,
        <div
          className=""
        >
          <WithStyles(DowColumn)
            dow="Mon"
            sections={Array []}
          />
          <WithStyles(DowColumn)
            dow="Tue"
            sections={Array []}
          />
          <WithStyles(DowColumn)
            dow="Wed"
            sections={Array []}
          />
          <WithStyles(DowColumn)
            dow="Thu"
            sections={Array []}
          />
          <WithStyles(DowColumn)
            dow="Fri"
            sections={Array []}
          />
        </div>,
      ],
      "className": "",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "hours": Array [
            8,
            9,
            10,
            11,
            12,
            13,
            14,
            15,
            16,
            17,
            18,
            19,
            20,
            21,
            22,
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <WithStyles(DowColumn)
              dow="Mon"
              sections={Array []}
            />,
            <WithStyles(DowColumn)
              dow="Tue"
              sections={Array []}
            />,
            <WithStyles(DowColumn)
              dow="Wed"
              sections={Array []}
            />,
            <WithStyles(DowColumn)
              dow="Thu"
              sections={Array []}
            />,
            <WithStyles(DowColumn)
              dow="Fri"
              sections={Array []}
            />,
          ],
          "className": "",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "Mon",
            "nodeType": "class",
            "props": Object {
              "dow": "Mon",
              "sections": Array [],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "Tue",
            "nodeType": "class",
            "props": Object {
              "dow": "Tue",
              "sections": Array [],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "Wed",
            "nodeType": "class",
            "props": Object {
              "dow": "Wed",
              "sections": Array [],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "Thu",
            "nodeType": "class",
            "props": Object {
              "dow": "Thu",
              "sections": Array [],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "Fri",
            "nodeType": "class",
            "props": Object {
              "dow": "Fri",
              "sections": Array [],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <WithStyles(HoursColumn)
            hours={
              Array [
                8,
                9,
                10,
                11,
                12,
                13,
                14,
                15,
                16,
                17,
                18,
                19,
                20,
                21,
                22,
              ]
            }
          />,
          <div
            className=""
          >
            <WithStyles(DowColumn)
              dow="Mon"
              sections={Array []}
            />
            <WithStyles(DowColumn)
              dow="Tue"
              sections={Array []}
            />
            <WithStyles(DowColumn)
              dow="Wed"
              sections={Array []}
            />
            <WithStyles(DowColumn)
              dow="Thu"
              sections={Array []}
            />
            <WithStyles(DowColumn)
              dow="Fri"
              sections={Array []}
            />
          </div>,
        ],
        "className": "",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "hours": Array [
              8,
              9,
              10,
              11,
              12,
              13,
              14,
              15,
              16,
              17,
              18,
              19,
              20,
              21,
              22,
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <WithStyles(DowColumn)
                dow="Mon"
                sections={Array []}
              />,
              <WithStyles(DowColumn)
                dow="Tue"
                sections={Array []}
              />,
              <WithStyles(DowColumn)
                dow="Wed"
                sections={Array []}
              />,
              <WithStyles(DowColumn)
                dow="Thu"
                sections={Array []}
              />,
              <WithStyles(DowColumn)
                dow="Fri"
                sections={Array []}
              />,
            ],
            "className": "",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "Mon",
              "nodeType": "class",
              "props": Object {
                "dow": "Mon",
                "sections": Array [],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "Tue",
              "nodeType": "class",
              "props": Object {
                "dow": "Tue",
                "sections": Array [],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "Wed",
              "nodeType": "class",
              "props": Object {
                "dow": "Wed",
                "sections": Array [],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "Thu",
              "nodeType": "class",
              "props": Object {
                "dow": "Thu",
                "sections": Array [],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "Fri",
              "nodeType": "class",
              "props": Object {
                "dow": "Fri",
                "sections": Array [],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
